### ISSUE TITLE
Problems including dynamic module synchronously

### ISSUE THREAD

**Issue by twoixter** [jspm/jspm-cli#1439]
I need to dynamically choose a module to include, but do it *synchronously*. I've tried using `System.import` but I don't know how to handle this as a promise. The use case is to import translations for **i18n**. I use `puttext` for translations and the problem is that the language is a dependency (I can't create an instance of the `puttext` object without first having the translations array).

The only way I can do this is using **AMD** or **CommonJS** modules, since ES6 import is static. (I've been doing this with RequireJS / r.js for a while).

I have some issues, which I list below. I don't know if these are real bugs or I'm doing something wrong. Please advice.

Take 1: As AMD module
--------------------------------

Sample code for `i18n.js` as AMD module. Edited for brevity.

```javascript
"format amd";

function get_document_lang()
{ return [ the locale code in documentElement.lang ] }

define(["puttext", "locale/" + get_document_lang() + "/LC_MESSAGES/interface.json!json"],
function(puttext, translations){
    return puttext(translations.locale_data.messages);
});
```

**Results**

* **Browser**: It works as expected.
* **CLI Bundling**: FAILS with the following error (edited for brevity): 

```
err  Error on instantiate for lib/i18n.js at file:///[...]/public/js/lib/i18n.js
	Loading app/main.js
	TypeError: Cannot read property 'processedValue' of undefined
         at /[...]/node_modules/systemjs-builder/compilers/amd.js:95:30
         at Array.map (native)
         at ParseTreeTransformer.AMDDependenciesTransformer.transformCallExpression (/[...]/node_modules/systemjs-builder/compilers/amd.js:94:38)
         at CallExpression.$__super.transform (/[...]/node_modules/traceur/bin/traceur.js:6623:28)
         at ParseTreeTransformer.transformAny (/[...]/node_modules/traceur/bin/traceur.js:9167:44)
         at ParseTreeTransformer.transformExpressionStatement (/[...]/node_modules/traceur/bin/traceur.js:9478:31)
[... several more lines ...]
```

Take 2: As CJS module
--------------------------------

Sample code for `i18n.js` as CJS module. Edited for brevity.

```javascript
"format cjs";

function get_document_lang()
{ return [ the locale code in documentElement.lang ] }

var puttx = require("puttext");
var translations = require("locale/" + get_document_lang() + "/LC_MESSAGES/interface.json!json");

exports.__ = puttx(translations.locale_data.messages);
```

**Results**

* **Browser**: FAILS with the following error in the console: [*Uncaught TypeError: Module locale/en_GB/LC_MESSAGES/interface.json!json not declared as a dependency*].
* **CLI Bundling**: Seems to compile, but when I ran it thru the browser I get the same error as the above.

Expectations
-----------------

* The above technique already works with RequireJS (and bundled with r.js), so I was expecting it to also work with SystemJS.

* **Take 1** (AMD format) *works in the browser* but fails at bundling time. I think it's due to the use of the call to `get_document_lang()` in the `define`. I've also tried with a simple variable and it fails with the same message. I suspect this to be a bug the AMD compiler **not expecting a expression** here since it also fails (same message) using `define(["locale/" + "en_US" + "/LC_MESSAGES/interface.json!json"]);`.

* **Take 2** (CJS format) might be not really an error, since reading the error message seems to suggest I can add all language modules to the config as dependencies of some other module.

**Comment by theefer** [comment]
I think you're right on both accounts; I'm not sure the AMD loader and bundler is expected to be able to cope with such semantics. It looks like the CJS one may be closer. Does making the "locale/..." dependency an explicitly relative one help?

```var translations = require("./locale/" + get_document_lang() + "/LC_MESSAGES/interface.json!json");```

**Comment by twoixter** [comment]
Hi @theefer , no, it also fails with same message. Thank you for the suggestion! I feel I'm close, but not there yet.

The AMD module is working without bundling, and it should work after accepting an expression at compiling time (so it can make it's `<script>` tag injection magic, whatever). This is how I have it working with RequireJS bundled with r.js.

I'm also having good feelings with the CJS version, though I'm unable to pass that *"not declared as dependency"* message. I've tried already with `System.set` and `System.registerDynamic` but to no avail.

```javascript
// Doesn't work
System.registerDynamic(["locale/" + _lang() + "/LC_MESSAGES/interface.json"], true, function(require, exports, module){
    var puttx = require("puttext");
    var trans = require("locale/" + _lang() + "/LC_MESSAGES/interface.json!json");
    module.exports = puttx(trans.locale_data.messages);
});
```

```javascript
// Doesn't work either
System.set(
    System.normalizeSync("locale/" + _lang() + "/LC_MESSAGES/interface.json!json"),
    System.newModule({})
);
```

**Comment by theefer** [comment]
We may have to invoke the wisdom of @guybedford at this point I'm afraid.

**Comment by guybedford** [comment]
SystemJS conditional interpolation is designed to handle this use case exactly. It's currently undocumented, but there's a section on it in the upcoming jspm beta guide. Perhaps have a look at that when it is out.

**Comment by twoixter** [comment]
@guybedford Thanks! Seems to be the way to go. I've been doing some tests with limited success (I've read the issues about conditional interpolation [here](https://github.com/systemjs/systemjs/issues/285)).

As noted in the issues, conditional builds seems not to work yet. This is what I get:

```javascript
// Module lib/i18n.js
import puttext from "puttext";
import translation from "locale/#{lib/browser-lang}/LC_MESSAGES/interface.json!json";

export default puttext(translation.locale_data.messages);
```

The module `lib/browser-lang.js` returns a string with the current locale, like `es_ES`.

This works as expected in the browser, but when I bundle, it fails with:

```
$ jspm bundle app/main

     Building the bundle tree for app/main...

err  TypeError: Cannot read property 'lastIndexOf' of undefined
         at getCanonicalName (./node_modules/systemjs-builder/lib/utils.js:110:32)
         at ./node_modules/systemjs-builder/lib/trace.js:282:55
         at Array.forEach (native)
         at ./node_modules/systemjs-builder/lib/trace.js:272:17
         at tryCatcher (./node_modules/systemjs-builder/node_modules/bluebird/js/release/util.js:11:23)
         at Promise._settlePromiseFromHandler (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:491:31)
         at Promise._settlePromise (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:548:18)
         at Promise._settlePromise0 (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:593:10)
         at Promise._settlePromises (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:676:18)
         at Promise._fulfill (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:617:18)
         at ./node_modules/systemjs-builder/node_modules/bluebird/js/release/nodeback.js:42:21
         at f (./node_modules/once/once.js:17:25)
         at Glob.<anonymous> (./node_modules/glob/glob.js:146:7)
         at emitOne (events.js:77:13)
         at Glob.emit (events.js:169:7)
         at Glob._finish (./node_modules/glob/glob.js:185:8)
```

I list this here in case it's useful. I don't know if it's more useful in the systemjs issue tracker.

I have a doubt though, what I'm supposing it to do when bundling is making this "require" for the interpolated URL dynamic so that I don't end up with several bundles one for each of `es_ES`, `en_GB`, `de_DE`, etc.

**Comment by guybedford** [comment]
@twoixter yes that's correct. I'm not sure it has been tested against plugin syntax yet actually. Try using metadata to set the json plugin, or use `meta: { '*.json': { format: 'json' }  }` rather and see if that resolves the build bug.

**Comment by twoixter** [comment]
@guybedford nope, it fails with same message of `Cannot read property...`. I'm using stable versions of jspm@0.16.24 and systems@0.19.17 now, I think I'll give up and wait new stable versions.

This is the config I've used anyway:

```javascript
System.config({
  [...]
  packages: {
      locale: {
          defaultExtension: false
      }
  },
  meta: {
      "*.json": {
          "loader": "json"
      }
  },
  [...]
});
```

**Comment by guybedford** [comment]
@twoixter it sounds like you're still using a `!` expression in your bundle command. This is a bug with plugin support for interpolation. I think I may have fixed it with https://github.com/systemjs/builder/commit/be382b2ea4cadc4df4f4272e8408e3f6431f560e if you're able to test that out, otherwise I'd suggest ensuring you're not using plugin syntax at all.

**Comment by twoixter** [comment]
@guybedford Nope, I'm pretty sure I deleted all `!` plugin expressions. The only instance is the one I mentioned for language loading.

In fact, if I change this line:

```javascript
import translation from "locale/#{lib/browser-lang}/LC_MESSAGES/interface.json";
```

to:

```javascript
import translation from "locale/en_GB/LC_MESSAGES/interface.json";
```

It does compile fine. As it probably should be, because it is now a regular static import. :-)

By the way, it also compiles if I add back the plugin:

```javascript
import translation from "locale/en_GB/LC_MESSAGES/interface.json!json";
```

I was tempted to dig a little into the error inside `systemjs-builder/lib/utils.js` but I'm afraid this is too deep into the rabbit hole for me. ;-)

(Using **jspm@0.16.24** and **systems@0.19.17**)

**Comment by guybedford** [comment]
@twoixter I mean specifically using plugin syntax in your bundle expression / task. The bug here is specifically about the interaction of plugin syntax and conditional interpolation in the build normalization system. Did you manage to try https://github.com/systemjs/builder/commit/be382b2ea4cadc4df4f4272e8408e3f6431f560e?

**Comment by twoixter** [comment]
@guybedford No, at least not that I'm aware of. :-) I'm now using this syntax in the `config.js` file:

```javascript
  [...]
  meta: {
      "*.json": {
          "loader": "json"
      }
  },
  [...]
```

I'm bundling with this command: `$ jspm bundle-sfx app/main`

Ok, I've had time now to try the patch of **systemjs/builder@be382b2** and it kind of worked! If a new error means an advance, here it is:

```sh
$ jspm bundle-sfx app/main
     Building the single-file sfx bundle for app/main...

err  TypeError: No resolution found at all for condition locale/#{lib/browser-lang.js}/LC_MESSAGES/interface.json.
         at ./node_modules/systemjs-builder/lib/trace.js:649:15
         at Array.forEach (native)
         at ./node_modules/systemjs-builder/lib/trace.js:642:6
         at tryCatcher (./node_modules/systemjs-builder/node_modules/bluebird/js/release/util.js:11:23)
         at Promise._settlePromiseFromHandler (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:491:31)
         at Promise._settlePromise (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:548:18)
         at Promise._settlePromise0 (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:593:10)
         at Promise._settlePromises (./node_modules/systemjs-builder/node_modules/bluebird/js/release/promise.js:676:18)
         at Async._drainQueue (./node_modules/systemjs-builder/node_modules/bluebird/js/release/async.js:125:16)
         at Async._drainQueues (./node_modules/systemjs-builder/node_modules/bluebird/js/release/async.js:135:10)
         at Immediate.Async.drainQueues [as _onImmediate] (./node_modules/systemjs-builder/node_modules/bluebird/js/release/async.js:16:14)
         at processImmediate [as _immediateCallback] (timers.js:383:17)
```

Reading this new error, I figured I must map somehow all possible combinations of the **lib/browser-lang.js** output (es_ES, en_US, de_DE, etc) but I'm stuck here.

**Comment by twoixter** [comment]
@guybedford Ok, an update on my previous comment.

Using `$ jspm bundle app/main` **without** sfx compiled fine and produced a working `build.js` suitable for production.

However, with this bundle.js I'm forced to use the following loader HTML:

```html
<script src="/js/jspm_packages/system.js"></script>
<script src="/js/config.js"></script>
<script src="/js/build.js"></script>
<script>System.import('app/main');</script>
```

Which is the documented [production workflow](https://github.com/jspm/jspm-cli/blob/master/docs/production-workflows.md). That's probably the desired behavior since **bundle-sfx** will try to statically build all available branches of the dynamic interpolation.

However, I'm still concerned with two issues:

* The above production bundle still makes XHR calls to load `/jspm_packages/github/systemjs/plugin-json@0.1.0.js` and `/jspm_packages/github/systemjs/plugin-json@0.1.0/json.js` to satisfy loader plugins. Which is weird, I expected those to be bundled with the main `bundle.js`.
* Even when injecting with `--inject` I think it's still mandatory to `<script>` the **config.js**. I expected to sfx all this to hide the complexities of **config.js** and remove all maps which are not required now since all dependencies are resolved.

**Comment by guybedford** [comment]
@twoixter it's really hard to follow all these interactions without spending much time on it. Is there any chance you can share a simple test case?

**Comment by twoixter** [comment]
@guybedford Yeah sorry, I agree. What would be the best way for me to add the test?

**Comment by guybedford** [comment]
Perhaps it would help to try and focus on specific questions. With regards to your last points:

> The above production bundle still makes XHR calls to load /jspm_packages/github/systemjs/plugin-json@0.1.0.js and /jspm_packages/github/systemjs/plugin-json@0.1.0/json.js to satisfy loader plugins. Which is weird, I expected those to be bundled with the main bundle.js.

This sounds like it would benefit from sharing your bundle file and configuration for me to see what is causing this. Alternatively you can also check why the names aren't matching up to see if `System.defined` contains the json plugin, or what in the bundle might be using the json plugin that isn't already inlined.

> It still bundles all variations of locale/{language}/LC_MESSAGES/* inside the bundle! Which is the main issue to avoid, I'm not willing to include all languages potentially adding megabytes of string texts. Although they are all included, I'm still seeing the dynamic XHR to load locale/{language}/LC_MESSAGES/interface.json as if the loader is unable to see them inside the bundle.

This is the default behaviour of conditional builds as described in http://jspm.io/0.17-beta-guide/conditional-interpolation.html. Conditional segments can be built by providing fixed conditions to bundle.

Again, that the bundle is not being used sounds like a configuration issue with the names not matching up which can be checked by verifying that the bundle names in the browser applied with paths and baseURL configuration match up exactly with the normalized import resolutions. Is anything loading from the bundle or is this only for json? What defaultExtension opt-out are you using? It's likely around that config assuming the baseURLs between build and client are matching up.

> Even when injecting with --inject I think it's still mandatory to <script> the config.js. I expected to sfx all this to hide the complexities of config.js and remove all maps which are not required now since all dependencies are resolved.

Bundles in jspm 0.16 are half-linked. In jspm 0.17 bundles are completely unlinked - that is they retain the need for configuration to be applied still in order to work. The reason for this is that the bundle is not a build, it is a multiplexing.

**Comment by twoixter** [comment]
@guybedford Thanks! Let do one thing, I think this diverged too much from the original issue, which was along the lines of "How to **synchronously** load something". I though it was easy, since Node and even RequireJS has synchronous semantics (`var foo = require("bar");` waits until foo has the contents of the "bar" module).

In order to close this issue, let me know if we just can continue over SystemJS (to help on the bundling issues, I need to check if it still fails with 0.17 bundler).

---

For completeness, this is the solution I came up to load something dynamically as dependency of my main application:

```html
<!-- main HTML loader -->
<script>
    System.import("./whatever" + whatever() + ".js").then(function(whatever){
        // Do something with "whatever", if applicable
        System.import("app/main");
    });
</script>
```

This way, the "app/main" bundle is not loaded until the required import is in place. I still think it would be a great feature if Jspm supported one (or both) ways of synchronously load something. See *Take 1* and *Take 2* of my first comment.

I've read the documentation of jspm 0.17-beta about interpolation, and although it is a great idea, it's far from ideal for translations for two reasons:

* It works for small projects, but is not ideal for large apps. Each language file of my app weights in about 300K. We have 9 languages now, looking to expand to 12 soon. It would amount near 4Mb of strings if we bundle all languages.
* Even if we do one bundle for each language, it is also not real-life, because we would need to deploy and bundle each time a translator changes a comma or fixes a translation (which happens often, go figure). Translation files should be decoupled from the main bundle, at least for our use case.

I initially though conditional interpolation would dynamically build an external request.

**Comment by guybedford** [comment]
The browser loader is asynchronous, there is no way it will ever be sync. I'd suggest reading the whole of http://jspm.io/0.17-beta-guide/conditional-interpolation.html, it describes how to create separate language bundles. Those bundles could themselves be subsets of bundle intersections via arithmetic.

**Comment by twoixter** [comment]
I'll give the conditional interpolation another try but as far I've seen, it doesn't quite fit our workflow.

Let me rephrase one last more time and we can close the issue, I promise. :-) By synchronously I mean the following:

```javascript
var content = require("./module.js");
console.log(content);
module.exports = ... whatever...
```

The above CommonJS snippet already works in jspm.io, and have "synchronous semantics" in the sense the second line does not get executed until "module.js" is loaded. I don't know the magic behind, I can imagine the snippet gets parsed and converted to some AMD format.

But the module name `"./module.js"` must be static. The following does not work:

```javascript
var content = require("./modu" + "le.js");  // Simple concatenation expression
console.log(content);
```

It throws an Exception like `Uncaught (in promise) Error: Module ./module.js not declared as a dependency.`. That is, it doesn't work if the parameter to `require` is an expression, not a string. It fails even though the expression gets evaluated, as noted by the exception message (it shows the full name of the module).

Do you think it would be possible to do this? Is it hard to implement?

**Comment by guybedford** [comment]
@twoixter did you read http://jspm.io/0.17-beta-guide/conditional-interpolation.html?
